text/3371-cargo-target-directories.md

```diff
@@ -0,0 +1,376 @@
+- Through the command line flag: `cargo build --target-directories /absolute/path/to/target/directories`
+
+The given path must be absolute: setting `CARGO_TARGET_DIRECTORIES` to an empty or relative path is an error (when used and not instantly overriden by `CARGO_TARGET_DIR`).
+
+## Resolution order relative to `CARGO_TARGET_DIR`
+
+The resolution order favors `CARGO_TARGET_DIR` in all its forms, in the interest of both backward compatibility and allowing overriding for a singular workspace:
+
+`--target-dir` > `CARGO_TARGET_DIR` > `target-dir = ...` > `--target-directories` > `CARGO_TARGET_DIRECTORIES` > `target-directories = ...`
+
+See the final section for discussion on how to make `CARGO_TARGET_DIRECTORIES` the default.
+
+## Naming
+
+In the example in the previous section, using `CARGO_TARGET_DIRECTORIES` with `cargo build` produces named subdirectories. The name of those is partially deterministic:
+it is the name of the parent directory of the workspace's `Cargo.toml` manifest and an unspecified hash of the absolute path to the workspace's root, so building `work-project/crate-1` will still use the `/cargo-caches/work-project-<hash>/debug/...` directory for a `cargo build` call.
+
+This naming scheme is chosen to be simple for people to navigate but is **not considered stable**: the hashing method (and so the hash) will probably not change often but `cargo` offers no guarantee and may change it in any release. Tools that needs to interact with `cargo`'s target directory should not rely on its value for more than a single invocation of them: they should instead query `cargo metadata` for the actual value each time they are invoked.
+
+In case the parent directory is `/` or `C:\`, the subdirectory name is implementation defined.
+
+The path used for the hashing and naming of the final directory is the one found *after* symlink resolution: `bazel` does it too and I have not found any complaints about this and it has the distinct advantage of allowing to make a symlink to a project somewhere else on the machine (for example to organise work projects) and avoid duplicating the build directory and all its data (which can be quite heavy).
+
+### Symbolic links
+
+In the following situation
+
+```
+/Users/
+├─ poliorcetics/
+│  ├─ projects/
+│  │  ├─ actual-crate/
+│  │  │  ├─ Cargo.toml
+│  │  ├─ symlink-to-crate/ -> actual-crate/
+```
+
+When calling `cargo metadata` in the `symlink-to-crate` path, the result contains `"manifest_path": "/Users/poliorcetics/projects/actual-crate/Cargo.toml"` and `"workspace_root":"/Users/poliorcetics/projects/actual-crate"`. This behaviour means that symlinks won't change the final directory used inside `CARGO_TARGET_DIRECTORIES`, or in other words: symbolic links are resolved.
+
+## Impact on `cargo ...` calls
+
+When calling `cargo` where `CARGO_TARGET_DIRECTORIES` is active, `CARGO_TARGET_DIR` is set by all `cargo` calls that happen in a Cargo workspace, including calls to third-party tools.
```

Comment:
> including calls to third-party tools.

This doesn't work for third-party cargo commands as this requires us to find the relevant `Cargo.toml` *but* it might be in the `--manifest-path` parameter passes to the third-party command which cargo never sees.

Reply:
ah thanks, fixed by saying to use the `cargo_metadata` crate in them